### PR TITLE
nasm: Fallback to native compiler when cross compiling

### PR DESCRIPTION
**Pull request by xclaesse**
If nasm is not defined in cross file binaries we can fallback to build machine nasm.

### CODE REVIEW

**Comment by eli-schwartz** [comment]
"we can" in the "this PR makes it do so" sense, or *we can" in the "this is supposed to work correctly" sense?

Why is this okay for nasm when it isn't okay for other languages whose compilers produce machine code?

**Comment by xclaesse** [comment]
> "we can" in the "this PR makes it do so" sense, or *we can" in the "this is supposed to work correctly" sense?

This PR makes Meson use nasm from build machine's PATH (or native file) for cross compiling, because there is no reason not to.

> Why is this okay for nasm when it isn't okay for other languages whose compilers produce machine code?

When cross compiling C code for Windows on Linux for example, you would typically use `gcc` for native targets and `x86_64-w64-mingw32-gcc` for cross targets. But for nasm code, the compiler is the same, you just pass `-fwin64`. Your source code has to be made for the target platform, not the compiler.

As far as I understand, at least.

**Comment by xclaesse** [comment]
CC @nirbheek who could have an opinion too.

**Comment by xclaesse** [comment]
> And does this flag automatically get added?

Yes, NasmCompiler class add the right arguments for the target platform.

> Does it matter which implementation you use (nasm, masm, metrowerks...)?

- masm is a different language, so it's not affected. I have no idea how cross compilation works with msvc/masm, if that's supported at all?
- metrowerks: I have no idea, I did not even know they provide nasm.

> Either way this needs to be explained in the commit message.

done

**Comment by eli-schwartz** [comment]
Thanks :) I'm on mobile or I'd answer some of those questions myself...

From a very quick look it looked like the detection method handled more than just nasm.exe but I could be misreading it.

**Comment by nirbheek** [comment]
> Why is this okay for nasm when it isn't okay for other languages whose compilers produce machine code?

Because gcc decided that you can't build a single binary that supports multiple targets. Not all compilers follow this paradigm.

rustc, for instance, supports everything via the same interface, just pass `--target <triple>`. nasm is the same: just pass the right `-f`.

yasm is the same as nasm. masm is not.

File: mesonbuild/compilers/detect.py [hunk 1]
```diff
@@ -1221,9 +1221,17 @@ def detect_swift_compiler(env: 'Environment', for_machine: MachineChoice) -> Com
 
 def detect_nasm_compiler(env: 'Environment', for_machine: MachineChoice) -> Compiler:
     from .asm import NasmCompiler, YasmCompiler, MetrowerksAsmCompilerARM, MetrowerksAsmCompilerEmbeddedPowerPC
-    compilers, _, _ = _get_compilers(env, 'nasm', for_machine)
     is_cross = env.is_cross_build(for_machine)
 
+    # When cross compiling and nasm is not defined in the cross file we can
+    # fallback to the build machine nasm.
```
Comment (nirbheek) [inline mesonbuild/compilers/detect.py:1227]: When cross-compiling, I think it makes sense to pick it up from the environment (`PATH` or `NASM`) but not from the native file. That would be surprising.

**Comment by xclaesse** [comment]
> rustc, for instance, supports everything via the same interface

I did not know that, I changed my implementation to support fallback to build machine compiler easier in the future.

**Comment by nirbheek** [comment]
@xclaesse could you please resolve the conflicts? Let's get this in (re: https://github.com/libass/libass/pull/330#issuecomment-2067625529).

**Comment by xclaesse** [comment]
@nirbheek rebased

**Comment by eli-schwartz** [comment]
Thanks for clarifying how this works and improving the commit message.


> > rustc, for instance, supports everything via the same interface
> 
> I did not know that, I changed my implementation to support fallback to build machine compiler easier in the future.

That's more of a clang thing than a rust thing. Anyways I think it makes sense to allow this to be generic the way you did, but not for either clang or rust -- rather, it's because there are other compilers that don't generate different outputs depending on host_machine.

As far as rust and clang go, we still want to enforce that people specify the correct compiler because there is no guarantee it works just because it is installed -- they still require having the "right" compiler (might mean installing a plugin for the target, which is essentially what the rust target components are) and the distinction as compared to gcc is actually just that the main compiler binary works as a wrapper driver *if* you have compiled it and/or installed nondefault extras to support the targets you want.

**Pull request merged**